### PR TITLE
Azure pipelines automation

### PR DESCRIPTION
**Pull request by dreycenfoiles**
This pull request contains a yaml file that will run a benchmarking process on a bi-weekly basis. The pipeline consists of running two shell scripts that will run the GPU and CPU focused processes in parallel.

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+- benchmarking
+
+jobs:
+
+- job: CPU_Benchmarks
+
+  pool:
+    name: Default
+    demands: 
+      - GPUs -equals 0
```
Comment (carterbox) [inline azure-pipelines.yml]: We talked about this a bit before, but what happens if we just let the Azure scheduler assign jobs? Isn't it smart enough to assign the GPU jobs appropriately without us using this demand-for-no-gpus?

File: azure-pipelines.yml [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+# Add steps that build, run tests, deploy, and more:
+# https://aka.ms/yaml
+
+schedules:
+  - cron: "0 0 */14 * *"
+    displayName: "Bi-monthly build"
+    branches:
+      include:
+        - master
+        - benchmarking
```
Comment (carterbox) [inline azure-pipelines.yml]:
```suggestion
```
We only need a cron job for the master branch I think.

File: azure-pipelines.yml [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      - Agent.OS -equals Linux
+  steps:
+
+  - script: |
+      conda env remove --yes -n tomopy-bench
+      rm -rf tomopy.github.io
+    displayName: Remove existing build environment
+
+  - bash: echo "##vso[task.prependpath]$CONDA/bin"
+    displayName: Add conda to PATH
```
Comment (carterbox) [inline azure-pipelines.yml]: It's weird that we are adding conda to the path after already using conda once. Swap these two steps?

File: azure-pipelines.yml [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+  steps: 
+
+  - script: |
+        conda env remove --yes -n tomopy-bench
+        rm -rf tomopy.github.io
+    displayName: Remove existing build environment
+
+  - bash: echo "##vso[task.prependpath]$CONDA/bin"
+    displayName: Add conda to PATH
```
Comment (carterbox) [inline azure-pipelines.yml]: Same comment as above.

**Review by carterbox — CHANGES_REQUESTED** [review]
`tomopy-bench.yaml` specifies `tomopy=1.2.1`, but GPU stuff wasn't implemented until `tomopy=1.5`! How is it possible that the GPU benchmarks are running!?

File: azure-pipelines.yml (hunk 1 above)
Comment (dreycenfoiles) [reply azure-pipelines.yml]: When I've tested out the scheduler, it will sometimes give the CPU-only job to one of the machines with a GPU. If we never plan on running more than two machines at a time, I guess there wouldn't be a problem then.

Comment (carterbox) [reply azure-pipelines.yml]: OK. In that case, I'd like a demand name that is more consistent across all of the workers. If there is a `-lessthan` and `-greaterthan` comparison for the demands, then add a capablility to all the workers called `GPU_COUNT` or something whose value is an integer. Otherwise, this demand should be `NO_GPU`.

Comment (dreycenfoiles) [reply azure-pipelines.yml]: There appears to be no support for inequality comparison.

Comment (carterbox) [reply azure-pipelines.yml]: :cry: Thanks for checking.

File: azure-pipelines.yml [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+      git config --global user.email "foilesdreycen@gmail.com"
+      git add *
+      git commit -m "Bi-monthly benchmark"
+      git push -f https://$(PAT)@github.com/tomopy/tomopy.github.io
+    displayName: 'Update benchmarking data repository'
+
+  - script: |
+      conda env remove --yes -n tomopy-benchmarks
+      rm -rf tomopy.github.io
+    displayName: Remove existing build environment
```
Comment (carterbox) [inline azure-pipelines.yml]: This step appears twice and is different each time.

File: azure-pipelines.yml [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+  - script: |
+      conda env remove --yes -n tomopy-bench
+      rm -rf tomopy.github.io
+    displayName: Remove existing build environment  
+
+  - script: |
+      conda env create -n tomopy-bench -f tomopy-bench.yaml 
+    displayName: 'Create benchmarking environment'
+
```
Comment (carterbox) [inline azure-pipelines.yml:38]:
```suggestion

    - script: conda list -n tomopy-bench
    displayName: 'List conda environment'

```

File: azure-pipelines.yml [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+      git config --global user.email "foilesdreycen@gmail.com"
+      git add *
+      git commit -m "Bi-monthly benchmark"
+      git push -f https://$(PAT)@github.com/tomopy/tomopy.github.io
+    displayName: 'Update benchmarking data repository'
+
+  - script: |
+      conda env remove --yes -n tomopy-benchmarks
+      rm -rf tomopy.github.io
+    displayName: Remove existing build environment
```
Comment (carterbox) [inline azure-pipelines.yml]: Again, this step is used twice, but is different.

File: azure-pipelines.yml [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+  - script: |
+        conda env remove --yes -n tomopy-bench
+        rm -rf tomopy.github.io
+    displayName: Remove existing build environment
+
+  - script: |
+      conda env create -n tomopy-bench -f tomopy-bench.yaml 
+    displayName: 'Create benchmarking environment'
+
```
Comment (carterbox) [inline azure-pipelines.yml:85]:
```suggestion
```suggestion

    - script: conda list -n tomopy-bench
    displayName: 'List conda environment'

```
```

File: tomopy-bench.yaml [hunk 9]
```diff
@@ -3,8 +3,9 @@ channels:
   - conda-forge
   - defaults
 dependencies:
+  - click
   - matplotlib=3
   - numpy=1.15
   - python=3.6
-  - tomopy=1.2.1
+  - tomopy=1.9=cuda10.2*
```
Comment (carterbox) [inline tomopy-bench.yaml]:
```suggestion
  - tomopy=*=cuda10.2*
```

File: benchmarking/summarize.py [hunk 10]
```diff
@@ -201,11 +202,14 @@ def image_quality_vs_time_plot(
     plt.xlim([0.1, 3600])
     plt.semilogx(basex=2)
     plt.xticks(
-        [0.1, 1, 5, 10, 30, 60, 5*60, 10*60, 30*60, 3600,],# 3*3600, 6*3600, 12*3600, 24*3600],
-        ['0.1s', '1s', '5s', '10s', '30s', '1m', '5m', '10m', '30m', '1h', '3h',],# '6h', '12h', '24h'],
+        [0.1, 1, 5, 10, 30, 60, 5*60, 10*60, 30*60,
+            3600, 3*3600, 6*3600, 12*3600, 24*3600],
+        ['0.1s', '1s', '5s', '10s', '30s', '1m', '5m',
+            '10m', '30m', '1h', '3h', '6h', '12h', '24h'],
```
Comment (carterbox) [inline benchmarking/summarize.py]: Please undo these changes.

File: benchmarking/summarize.py [hunk 11]
```diff
@@ -218,14 +222,15 @@ def image_quality_vs_time_plot(
 
 def concat_recons(algo_folder, base_path):
     recons = list()
-    for file in glob.glob(os.path.join(algo_folder, "*.jpg")):
+    for file in glob.glob(os.path.join(algo_folder, "*.png")):
```
Comment (carterbox) [inline benchmarking/summarize.py]: These changes are not within the scope of this PR.

**Review by carterbox — CHANGES_REQUESTED** [review]

File: envs/tomopy-bench-gpu.yaml [hunk 12]
```diff
@@ -0,0 +1,11 @@
+name: tomopy-cpu
```
Comment (carterbox) [inline envs/tomopy-bench-gpu.yaml:1]:
```suggestion
name: tomopy-gpu
```

**Review by carterbox — APPROVED** [review]
